test/Constraints/generics.swift

```diff
@@ -592,3 +596,28 @@ func sr8075() {
 
   let _: UIFont = .init(ofSize: switchOnCategory([0: 15.5, 1: 20.5]))
 }
+
+// rdar://problem/40537858 - Ambiguous diagnostic when type is missing conformance
+func rdar40537858() {
+  struct S {
+    struct Id {}
+    var id: Id
+  }
+
+  struct List<T: Collection, E: Hashable> {
+    typealias Data = T.Element
+    init(_: T, id: KeyPath<Data, E>) {}
+  }
+
+  var arr: [S] = []
+  _ = List(arr, id: \.id) // expected-error {{'List<[S], S.Id>' requires that 'S.Id' conform to 'Hashable'}}
```

Comment:
Eventually, we should say something like

```
List<T, E> requires that 'E' conform to 'Hashable' (with 'E' = 'S.Id')

Reply:
I can definitely add requirement information to the diagnostic message, since owner is known I can get generic signature and extract the original type from requirement.

Reply:
Yeah. Including the name of the decl and it’s generic signature would make the diagnostic clearer, unless it becomes too long; but in that case we can split off some notes to spell out the substitutions, for example